pilot/pkg/model/network.go

```diff
@@ -0,0 +1,144 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"net"
+)
+
+const (
+	noCluster = ""
+)
+
+// NetworkID is the unique identifier for a network.
+type NetworkID string
+
+func (id NetworkID) Equals(other NetworkID) bool {
+	return SameOrEmpty(string(id), string(other))
+}
+
+// ClusterID is the unique identifier for a k8s cluster.
+type ClusterID string
+
+func (id ClusterID) Equals(other ClusterID) bool {
+	return SameOrEmpty(string(id), string(other))
+}
+
+// Gateway is the gateway of a network
+type Gateway struct {
+	// Network is the ID of the network where this Gateway resides.
+	Network NetworkID
+	// Cluster is the ID of the k8s cluster where this Gateway resides.
+	Cluster ClusterID
+	// gateway ip address
+	Addr string
+	// gateway port
+	Port uint32
+}
+
+// Gateways provides information about network Gateway resources throughout the mesh.
+type Gateways interface {
+	// IsMultiNetworkEnabled returns true if one or more Gateways are configured.
+	IsMultiNetworkEnabled() bool
+
+	// All returns all configured Gateway resources in the entire mesh.
+	All() []*Gateway
+
+	// ForNetwork returns all Gateway resources configured for the given network.
+	ForNetwork(network NetworkID) []*Gateway
+
+	// ForNetworkAndCluster returns all Gateway resources configured for the given network and cluster.
+	ForNetworkAndCluster(network NetworkID, cluster ClusterID) []*Gateway
+}
+
+// newGateways creates a new Gateways from the Environment by merging
+// together the MeshNetworks and ServiceRegistry-specific gateways.
+func newGateways(env *Environment) Gateways {
+	// Generate the a snapshot of the state of gateways by merging the contents of
+	// MeshNetworks and the ServiceRegistries.
+	gatewayMap := make(map[NetworkID]map[ClusterID][]*Gateway)
+
+	addGateway := func(gateway *Gateway) {
```

Comment:
nit: why not make this a method on `gatewaysImpl` – the only thing we're accessing in the parent scope is `gatewayMap` which is essentially `gatewaysImpl.gateways`

Reply:
This seemed more appropriate, since `Gateways` is effectively immutable. Adding an `addGateway` method would suggest that it isn't.